Handle async SSLHandshake properly

DarrenTsung opened the pull request:
Handle a WouldBlock error returned from Ssl::connect() properly by storing the MidHandshake returned and using it in the OpensslStream until handshake() is established.

Tested with `RUST_LOG=hyper=debug cargo run --example client https://httpbin.org/status/418` to examine the internal calls. Also tested with `hyper-client-pool` (https://github.com/onesignal/hyper-client-pool) tests which were failing after update to openssl 0.10.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jwilm/hyper/2)
<!-- Reviewable:end -->

Commit:
Add backtrace

Commit:
Change to debug

Commit:
Switch to fork with 0.7 connect functionality

Commit:
Use direct git instead of replace

Commit:
Openssl back to 0.10

Commit:
Handle WouldBlock when handshake correctly

Commit:
Use macro_use only in tests

Commit:
Handle new WANT_READ / WANT_WRITE possibilities with handshake

Commit:
Remove backtrace dependency

Commit:
Remove unused use AsciiExt

jwilm commented:
I tested this and it works great. Have some questions in the code.

---

Reviewed 8 of 9 files at r1.
Review status: 8 of 9 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/net.rs, line 438 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BuOBpVIu8Q0coBXux:-L5BuOBpVIu8Q0coBXuy:b-tugbu2) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L438)):*
> ```Rust
>     use std::cell::Cell;
>     use std::path::Path;
>     use std::rc::Rc;
> ```

I don't think this is needed. I'll take a shot at removing it.

---

*[src/net.rs, line 568 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BxJRaKy6jlT_BtHLJ:-L5BxJRbB2okPfaNPOPa:b-kqwgr7) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L568)):*
> ```Rust
> 
>     #[derive(Debug)]
>     enum OpensslStreamInner<T> where T: Debug {
> ```

Why is the `Debug` bound needed?

---

*[src/net.rs, line 616 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BunNldJvvLdhITcuj:-L5BunNldJvvLdhITcuk:bd79arj) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L616)):*
> ```Rust
>             let blocked = self.blocked.clone();
> 
>             loop {
> ```

Why is this a loop? It looks like every possible branch has a return value.

---

*[src/net.rs, line 622 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BwIQ4HHvlnoFsqbKU:-L5BwIQ4HHvlnoFsqbKV:btjxqzg) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L622)):*
> ```Rust
>                             Ok(stream) => {
>                                 debug!("read - Resolved handshake, connected stream!");
>                                 self.inner = Some(OpensslStreamInner::Connected(stream));
> ```

It seems to me that this could recurse when changing types so that it doesn't need another spin through the event loop.

---

*[src/net.rs, line 628 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BwWC4eC26T-JJXjIq:-L5BwWC4eC26T-JJXjIr:b-jcennh) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L628)):*
> ```Rust
>                                 debug!("read - Handshake would block!");
>                                 self.inner = Some(OpensslStreamInner::Connecting(stream));
>                                 blocked.set(Some(Blocked::Write));
> ```

How do you know it's blocked on a *write*, specifically?

---

*[src/net.rs, line 671 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BwlFaq6LXmzqPL-IF:-L5BwlFaq6LXmzqPL-IG:b9htb87) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L671)):*
> ```Rust
>             let blocked = self.blocked.clone();
> 
>             loop {
> ```

Ditto on the loop here.

---

*[src/net.rs, line 702 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5Bx1ReRdZlktWCa35b:-L5Bx1Rf1ON4tpbB4VbR:bdz96pv) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L702)):*
> ```Rust
>                                 Err(e)
>                             },
>                             ErrorCode::WANT_WRITE => {
> ```

I don't think this needs to be handled in `write`. At least, it wasn't being done earlier.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jwilm/hyper/2#-:-L5ByJj_tfdhE88eBJQx:bczjwf6)*
<!-- Sent from Reviewable.io -->

jwilm committed:
Remove unnecessary Rc

jwilm commented:
Reviewed 1 of 9 files at r1.
Review status: 8 of 9 files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/jwilm/hyper/2)*
<!-- Sent from Reviewable.io -->

jwilm committed:
Cleanup

- Removes unnecessary `Debug` bounds
- Refactors _handshake_ handling into reusable function
- Fixes match references
- Make error handling more like it was originally

jwilm commented:
I made some changes and left a couple more comments. I'm not marking my changes as reviewed so you can have a chance to review as well.

---

Review status: 8 of 9 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/net.rs, line 568 at r1](https://beta.reviewable.io/reviews/jwilm/hyper/2#-L5BxJRaKy6jlT_BtHLJ:-L5CVGZd7MWBUuwgSHWu:b-hncw67) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L568)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Why is the `Debug` bound needed?
</blockquote></details>

I removed this where it wasn't needed.

---

*[src/net.rs, line 616 at r1](https://beta.reviewable.io/reviews/jwilm/hyper/2#-L5BunNldJvvLdhITcuj:-L5CVJaa1x70miZzKXiD:b-ne281m) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L616)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Why is this a loop? It looks like every possible branch has a return value.
</blockquote></details>

I removed the loop since we weren't using it.

---

*[src/net.rs, line 622 at r1](https://beta.reviewable.io/reviews/jwilm/hyper/2#-L5BwIQ4HHvlnoFsqbKU:-L5CVQv_CxwsaT8VlTgh:b-6pgjoe) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L622)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

It seems to me that this could recurse when changing types so that it doesn't need another spin through the event loop.
</blockquote></details>

My initial comment here isn't correct, but I'm also worried that we could "get stuck" with this socket if we return WouldBlock here when there's data available (pretty sure rotor is edge-triggered).

---

*[src/net.rs, line 628 at r1](https://beta.reviewable.io/reviews/jwilm/hyper/2#-L5BwWC4eC26T-JJXjIq:-L5CVfA4Cf3qOqrqiIUW:b267tbw) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L628)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

How do you know it's blocked on a *write*, specifically?
</blockquote></details>

I updated this to check the stream error code so that it would guarantee requesting the correct operation.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/jwilm/hyper/2#-:-L5CVrZ01Z9_V9YTeg_Y:boug6ci)*
<!-- Sent from Reviewable.io -->

DarrenTsung commented:
Review status: 8 of 9 files reviewed at latest revision, 7 unresolved discussions.

---

*[src/net.rs, line 438 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BuOBpVIu8Q0coBXux:-L5F6jJv-ut_nhVvEsbb:b-9triw2) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L438)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

I don't think this is needed. I'll take a shot at removing it.
</blockquote></details>

Ah okay thanks for the cleanup! 

---

*[src/net.rs, line 568 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BxJRaKy6jlT_BtHLJ:-L5F7PAgq1q6kF1kCCvI:b-9755fn) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L568)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

I removed this where it wasn't needed.
</blockquote></details>

Thank you!

---

*[src/net.rs, line 616 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BunNldJvvLdhITcuj:-L5F7QLyMb_KwtKipsQV:b6g1pll) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L616)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

I removed the loop since we weren't using it.
</blockquote></details>

Thanks! I had the idea that you had where it should just recurse after connecting, but it ended up not working.

---

*[src/net.rs, line 628 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BwWC4eC26T-JJXjIq:-L5F7xAOQCXlNmJEGs8P:b-8cih0) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L628)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

I updated this to check the stream error code so that it would guarantee requesting the correct operation.
</blockquote></details>

Thank you, good point!

---

*[src/net.rs, line 671 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BwlFaq6LXmzqPL-IF:-L5F7ziyGfwZv1-laamx:b2nhtan) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L671)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Ditto on the loop here.
</blockquote></details>

Yup

---

*[src/net.rs, line 702 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5Bx1ReRdZlktWCa35b:-L5FEoPoL-C2Sxf5DW-o:bm1na94) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L702)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

I don't think this needs to be handled in `write`. At least, it wasn't being done earlier.
</blockquote></details>

Yeah this is a change I made after examining the tests in hyper-client-pool and seeing that they don't resolve the connection successfully. I will try to implement a test that fails this case in hyper.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jwilm/hyper/2)*
<!-- Sent from Reviewable.io -->

Commit:
Add failing test from hyper-client-pool

Commit:
Fix failing test by handling block on same WRITE / READ

jwilm commented:
Review status: 8 of 10 files reviewed at latest revision, 3 unresolved discussions.

---

*[src/net.rs, line 616 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BunNldJvvLdhITcuj:-L5FHyzfzRbb3o7LDEtY:b8jt30s) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L616)):*
<details><summary><i>Previously, DarrenTsung (Darren Tsung) wrote…</i></summary><blockquote>

Thanks! I had the idea that you had where it should just recurse after connecting, but it ended up not working.
</blockquote></details>

Yes it would be nice to recurse. However, I don't think we're in a position to do so. Handshake first requires write, then read. A client's next action is to write, so the read will certainly return not ready, and since we're in `read`, there's not a buffer provided to `write`.

I mentioned in another comment that I'm concerned about "getting stuck" due to edge triggered behavior. I don't know of a clean way to address this, though. Perhaps we need to look at how this behaved in 0.7 regarding read/write calls and socket interest.

---

*[src/net.rs, line 702 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5Bx1ReRdZlktWCa35b:-L5FHeqskywVpMAY4_UN:b-ikok4t) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L702)):*
<details><summary><i>Previously, DarrenTsung (Darren Tsung) wrote…</i></summary><blockquote>

Yeah this is a change I made after examining the tests in hyper-client-pool and seeing that they don't resolve the connection successfully. I will try to implement a test that fails this case in hyper.
</blockquote></details>

Did my changes reintroduce a bug in hyper-client-pool?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jwilm/hyper/2)*
<!-- Sent from Reviewable.io -->

DarrenTsung commented:
Review status: 8 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/net.rs, line 702 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5Bx1ReRdZlktWCa35b:-L5FIVYqsXr92vaVQ47B:b8txyl4) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L702)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Did my changes reintroduce a bug in hyper-client-pool?
</blockquote></details>

Yeah something is different between the hyper tests + hyper-client-pool tests. I added in a new test pulled from hyper-client-pool that fails.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jwilm/hyper/2)*
<!-- Sent from Reviewable.io -->

DarrenTsung commented:
Review status: 8 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/net.rs, line 616 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BunNldJvvLdhITcuj:-L5FJ-mRc8kpMhlGSm-v:b-upll46) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L616)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

Yes it would be nice to recurse. However, I don't think we're in a position to do so. Handshake first requires write, then read. A client's next action is to write, so the read will certainly return not ready, and since we're in `read`, there's not a buffer provided to `write`.

I mentioned in another comment that I'm concerned about "getting stuck" due to edge triggered behavior. I don't know of a clean way to address this, though. Perhaps we need to look at how this behaved in 0.7 regarding read/write calls and socket interest.
</blockquote></details>

Yep, that is what happened. I can take a look at openssl 0.7, but might sync up with you to discuss if I'm looking at the right stuff :P.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jwilm/hyper/2)*
<!-- Sent from Reviewable.io -->

DarrenTsung commented:
Review status: 8 of 10 files reviewed at latest revision, 2 unresolved discussions.

---

*[src/net.rs, line 622 at r1](https://reviewable.io:443/reviews/jwilm/hyper/2#-L5BwIQ4HHvlnoFsqbKU:-L5FJVVteF6NSvlsm4DG:bfosyml) ([raw file](https://github.com/jwilm/hyper/blob/0dffe36c3609165dd4e1777937b8b06f4479d287/src/net.rs#L622)):*
<details><summary><i>Previously, jwilm (Joe Wilm) wrote…</i></summary><blockquote>

My initial comment here isn't correct, but I'm also worried that we could "get stuck" with this socket if we return WouldBlock here when there's data available (pretty sure rotor is edge-triggered).
</blockquote></details>

^taking a look at openssl 0.7, will sync up with you in a bit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/jwilm/hyper/2)*
<!-- Sent from Reviewable.io -->
